src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/MarkMembersAsStatic.cs

```diff
@@ -91,57 +93,61 @@ void OnMethodReference(OperationAnalysisContext operationContext)
 
                 void OnOperationBlockStart(OperationBlockStartAnalysisContext blockStartContext)
                 {
-                    if (blockStartContext.OwningSymbol is not IMethodSymbol methodSymbol)
-                    {
-                        return;
-                    }
-
-                    // Don't run any other check for this method if it isn't a valid analysis context
-                    if (!ShouldAnalyze(methodSymbol, wellKnownTypeProvider, skippedAttributes, isWebProject, blockStartContext))
+                    if (blockStartContext.OwningSymbol is not IMethodSymbol methodSymbol ||
+                        // Don't run any other check for this method if it isn't a valid analysis context
+                        !ShouldAnalyze(methodSymbol, wellKnownTypeProvider, skippedAttributes, isWebProject, blockStartContext))
                     {
                         return;
                     }
 
+                    var localVariables = PooledConcurrentSet<ISymbol>.GetInstance();
                     bool isInstanceReferenced = false;
 
                     blockStartContext.RegisterOperationAction(operationContext =>
-                    {
-                        if (((IInstanceReferenceOperation)operationContext.Operation).ReferenceKind == InstanceReferenceKind.ContainingTypeInstance)
-                        {
-                            isInstanceReferenced = true;
-                        }
-                    }, OperationKind.InstanceReference);
+                        CheckHasInstanceReference(operationContext.Operation, ref isInstanceReferenced)
+                        , OperationKind.InstanceReference);
 
-                    // Workaround for https://github.com/dotnet/roslyn/issues/27564
                     blockStartContext.RegisterOperationAction(operationContext =>
+                        CheckHasReferenceInXml(operationContext.Operation, ref isInstanceReferenced)
+                        , OperationKind.None);
+
+                    if (SupportStaticLocalFunctions(blockStartContext.OperationBlocks[0].Syntax.SyntaxTree.Options))
```

Comment:
Maybe this check should be done at the symbol level instead of at each block.

Reply:
@Evangelink It actually can be done at compilation start, then pass the boolean value around

Reply:
Yes, I was just wondering if it could cause some problem with partial definitions.

Reply:
@Evangelink Partial definitions should never be analyzed, only partial implementations does. https://github.com/dotnet/roslyn-analyzers/pull/5172